script/cache_critical_dns

```diff
@@ -170,6 +170,9 @@ class SRVName
       [].tap do |addresses|
         targets = dns_client.getresources(@name, Resolv::DNS::Resource::IN::SRV)
         targets.delete_if { |t| !@prio_filter.within_threshold?(t.priority) }
+        # Targets are resolved in reverse priority order, so they are
+        # considered the first_seen in the ResolverCache entry.
+        targets.sort_by!(&:priority).reverse!
```

Comment:
I found this very hard to understand.  You seem to implicitly assume unique values for `Time.now` around each iteration of the loop in `ResolverCache`.

```
@name.resolve.each do |address|
  if @cached[address]
    @cached[address].last_seen = Time.now.utc
  else
    @cached[address] = CacheMeta.new(Time.now.utc, Time.now.utc)
  end
end
```

I don't think this is a reasonable assumption.  (Maybe it holds today, in MRI and on our kernels, but I would not bank on it.)

`ResolverCache` should be solely responsible for all ordering.  We should supply it with enough data to make a sensible ordering decision.

Could we return priority values from `Name.resolve` and `SRVName.resolve`?

```
-> [(prio, addr), ...]
```

`ResolverCache` could then trivially order by `prio` then break ties with its internal `first_seen` time.

As this would alter the data type of the value returned by `Name.resolve` and `SRVName.resolve`, I would change the name of the methods to `resolve_with_priority`.

Alternatively -- and this might be better -- you could keep `Name.resolve` as it is (no priorities here), and use a `respond_to?(:resolve_with_priority)` test in `ResolverCache.resolve`.

Separately, I think there was another other bug here.  `SRVName.resolve` aborts too early on a DNS exception.  Suppose it finds three targets, but only one fails to resolve (to a v4/6 address).  I think it should return a usable result for the other two targets.

Here is my suggestion as a patch.
Usual disclaimer:  I don't know ruby, and I have not tested. :)

```diff
diff --git a/script/cache_critical_dns b/script/cache_critical_dns
index 803b0216e6..dc4bbf3207 100755
--- a/script/cache_critical_dns
+++ b/script/cache_critical_dns
@@ -166,17 +166,26 @@ class SRVName
   end
 
   def resolve
+    resolve_with_priority.map(&:last).uniq
+  end
+
+  def resolve_with_priority
     dns_client_with_timeout do |dns_client|
-      [].tap do |addresses|
+      [].tap do |results|
         targets = dns_client.getresources(@name, Resolv::DNS::Resource::IN::SRV)
         targets.delete_if { |t| !@prio_filter.within_threshold?(t.priority) }
-        addresses.concat(targets.map { |t| Name.new(t.target.to_s).resolve }.flatten)
+        targets.each do |t|
+          Name.new(t.target.to_s).resolve.each do |addr|
+            results << [t.priority, addr]
+          rescue Resolv::ResolvError, Resolv::ResolvTimeout
+          end
+        end
       end
     end
   end
 end
 
-CacheMeta = Struct.new(:first_seen, :last_seen)
+CacheMeta = Struct.new(:priority, :first_seen, :last_seen)
 
 class EmptyCacheError < StandardError; end
 
@@ -189,30 +198,51 @@ class ResolverCache
     @cached = {}
   end
 
-  # Returns a list of resolved addresses ordered by first seen time.  Most
-  # recently seen address is first.
-  # If an exception occurs during DNS resolution we return whatever addresses
-  # are cached.
+  # Resolve @name and return an array of resolved addresses.
+  # Addresses are ordered by priority and time they were first seen.
+  # Addresses with lower numeric priority are always returned first.
+  # Given more than one address of equal priority,
+  # the address that was most recently seen is returned first.
+  # Cached addresses are returned upon any DNS exception.
   # Addresses last seen more than 30 minutes ago are evicted from the cache.
-  # Raises EmptyCacheError if the cache is empty after DNS resolution and cache
-  # eviction is performed.
+  # Raises EmptyCacheError if the cache is found to be empty
+  # after DNS resolution and cache eviction.
   def resolve
-    begin
-      @name.resolve.each do |address|
-        if @cached[address]
-          @cached[address].last_seen = Time.now.utc
-        else
-          @cached[address] = CacheMeta.new(Time.now.utc, Time.now.utc)
-        end
+    new_results = begin
+      if @name.respond_to?(:resolve_with_priority)
+        @name.resolve_with_priority
+      else
+        # synthesise a priority
+        # literal 0 is not significant:  any number will do
+        @name.resolve.map { |addr| [0, addr] }
       end
     rescue Resolv::ResolvError, Resolv::ResolvTimeout
+      []
     end
 
-    @cached = @cached.delete_if { |_, meta| Time.now.utc - 30 * 60 > meta.last_seen }
+    now = Time.now.utc
+    new_results.each do |r|
+      prio, addr = r
+      if @cached[addr]
+        @cached[addr].priority  = prio
+        @cached[addr].last_seen = now
+      else
+        @cached[addr] = CacheMeta.new(prio, now, now)
+      end
+    end
+
+    @cached = @cached.delete_if { |_, meta| now - 30 * 60 > meta.last_seen }
+
     if @cached.empty?
       raise EmptyCacheError, "DNS resolver found no usable addresses"
     end
-    @cached.sort_by { |_, meta| meta.first_seen }.reverse.map(&:first)
+
+    @cached.sort_by do |_, meta|
+      [
+        meta.priority,
+        now - meta.first_seen,
+      ]
+    done.map(&:first).uniq
   end
 end
```

Reply:
```
results << [t.priority, addr]
```

Sorry, on second thought, maybe it would be better to swap this around (in `resolve_with_priority`).

```
results << [addr, t.priority]
```

There would be no functional difference (once callers are similarly adjusted), though I think this would be a little less smelly:  if we call a method whose name begins with `resolve`, it is probably the addresses we most care about. :)

Anyway, up to you.